File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void backtrack(string digits, vector<string>&letter ,vector<string>&ans, int index, string str){
        //base condition
        if(index == digits.size()){
            ans.push_back(str);
            return;
        }
        string possibleLetter = letter[digits[index]-'2'];
        for(int i=0; i<possibleLetter.size(); i++){
            str.push_back(possibleLetter[i]);
            backtrack(digits, letter, ans, index+1, str);
            str.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        vector<string>letter = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string>ans;
        if(digits == "") return ans;
        backtrack(digits, letter, ans, 0, "");
        return ans;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    List<String>ans;
    void helper(int index, List<String>list, StringBuilder current){
        if(index>=list.size()){
            ans.add(new String(current.toString()));
            return;
        }
        String str = list.get(index);
        for(int i=0; i<str.length(); i++){
            current.append(str.charAt(i));
            helper(index+1, list, current);
            current.deleteCharAt(current.length()-1);
        }
    }
    public List<String> letterCombinations(String digits) {
        String[] str = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        List<String>list = new ArrayList<>();
        for(char ch: digits.toCharArray()){
            list.add(str[ch-'2']);
        }
        ans = new ArrayList<>();
        helper(0, list, new StringBuilder());
        return ans;
    }
}